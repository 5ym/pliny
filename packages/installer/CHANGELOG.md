# @pliny/installer

## 0.0.1-rc.8

### Patch Changes

- d78d04f: create separate tsconfig to resolve recipes

## 0.0.1-rc.7

### Patch Changes

- cd6305d: add missing react dependency for installer

## 0.0.1-rc.6

### Patch Changes

- 5d0eb8d: run prettier only on target directory

## 0.0.1-rc.5

### Patch Changes

- 264a96b: remove next/compiled and add packages as dependency
- 5dbf7d1: remove build-publishable-release script

## 0.0.1-rc.4

### Patch Changes

- c78af92: bump all packages

## 0.0.1-rc.3

### Patch Changes

- 11e43cc: patch release for changelog to work with workspace:\*
