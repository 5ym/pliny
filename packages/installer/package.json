{
  "name": "@pliny/installer",
  "version": "0.0.1",
  "description": "Package installation for the Pliny CLI",
  "homepage": "https://github.com/timlrx/pliny",
  "license": "MIT",
  "scripts": {
    "dev": "tsup-node src/* --watch --no-splitting --format esm,cjs --dts --silent",
    "build": "rimraf dist && tsup-node src/* --no-splitting --format esm,cjs --dts --silent"
  },
  "author": "Timothy Lin <timothy0336@hotmail.com> (https://timlrx.com)",
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "keywords": [
    "pliny",
    "installer"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/timlrx/pliny.git"
  },
  "dependencies": {
    "@babel/core": "7.18.0",
    "@babel/plugin-transform-typescript": "7.18.1",
    "@types/jscodeshift": "0.11.5",
    "ast-types": "0.14.2",
    "console-table-printer": "^2.7.5",
    "cross-spawn": "7.0.3",
    "diff": "5.0.0",
    "enquirer": "2.3.6",
    "execa": "6.1.0",
    "fs-extra": "^9.1.0",
    "globby": "11.0.2",
    "ink": "3.2.0",
    "ink-spinner": "4.0.3",
    "ink-testing-library": "2.1.0",
    "jscodeshift": "0.13.1",
    "mem-fs": "2.2.0",
    "mem-fs-editor": "9.4.0",
    "ora": "5.4.1",
    "pluralize": "^8.0.0",
    "react": "17.0.2",
    "recast": "0.20.5",
    "rimraf": "^3.0.2",
    "tslog": "^3.1.1",
    "vinyl": "2.2.1"
  },
  "devDependencies": {
    "tsup": "5.12.0",
    "typescript": "^4.6.2",
    "contentlayer": "0.2.5"
  },
  "publishConfig": {
    "access": "public"
  }
}
