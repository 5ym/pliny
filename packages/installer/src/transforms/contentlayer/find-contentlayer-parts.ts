import j from 'jscodeshift'
import { Program } from '../../types'

export const findImports = (program: Program) => {
  return program.find(j.ImportDeclaration).paths()
}

export const findMakeSource = (program: Program) => {
  // In contentlayer make source should always be default exported
  return program.find(j.ExportDefaultDeclaration).paths()
}

export const findContentlayerDefineDocument = (program: Program) => {
  const defineDocument = program.find(j.VariableDeclaration).filter((path) => {
    const firstInnerDeclaration = path.value.declarations[0]
    let status = false
    // path.name is a number if it contains an array i.e. a body node
    if (typeof path.name === 'number' && firstInnerDeclaration.type == 'VariableDeclarator') {
      const init = firstInnerDeclaration.init
      if (init.type == 'CallExpression') {
        const callee = init.callee
        if (callee.type == 'Identifier' && callee.name == 'defineDocumentType') {
          status = true
        }
      }
    }
    return status
  })
  const exportedDefineDocument = program.find(j.ExportNamedDeclaration).filter((path) => {
    const declaration = path.value.declaration
    let status = false
    if (declaration.type == 'VariableDeclaration' && declaration.declarations) {
      const firstInnerDeclaration = declaration.declarations[0]
      if (firstInnerDeclaration.type == 'VariableDeclarator') {
        const init = firstInnerDeclaration.init
        if (init.type == 'CallExpression') {
          const callee = init.callee
          if (callee.type == 'Identifier' && callee.name == 'defineDocumentType') {
            status = true
          }
        }
      }
    }
    return status
  })
  const allPaths = [...defineDocument.paths(), ...exportedDefineDocument.paths()]
  return allPaths
}
