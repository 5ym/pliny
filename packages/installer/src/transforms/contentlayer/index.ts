export * from './find-contentlayer-parts'
