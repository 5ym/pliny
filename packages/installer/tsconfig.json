{
  "extends": "@pliny/config/tsconfig.base.json",
  "compilerOptions": {
    "rootDir": ".",
    "baseUrl": "./",
    "outDir": "lib"
  },
  "include": ["src", "types"]
}
