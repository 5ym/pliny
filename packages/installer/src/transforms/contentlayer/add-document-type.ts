import type { ExpressionKind } from 'ast-types/gen/kinds'
import j from 'jscodeshift'
import { JsonValue, Program } from '../../types'
import { findContentlayerParts } from './find-contentlayer-parts'

const jsonValueToExpression = (value: JsonValue): ExpressionKind =>
  typeof value === 'string'
    ? j.stringLiteral(value)
    : typeof value === 'number'
    ? j.numericLiteral(value)
    : typeof value === 'boolean'
    ? j.booleanLiteral(value)
    : value === null
    ? j.nullLiteral()
    : Array.isArray(value)
    ? j.arrayExpression(value.map(jsonValueToExpression))
    : j.objectExpression(
        Object.entries(value)
          .filter((entry): entry is [string, JsonValue] => entry[1] !== undefined)
          .map(([key, value]) =>
            j.objectProperty(j.stringLiteral(key), jsonValueToExpression(value))
          )
      )

export function addDocumentType(
  program: Program,
  variableName: string,
  documentTypeDef: JsonValue
): Program {
  const { imports, defineDocument, makeSource } = findContentlayerParts(program)

  const ast = j.variableDeclaration('const', [
    j.variableDeclarator(
      j.identifier(variableName),
      j.callExpression(j.identifier('defineDocumentType'), [
        j.arrowFunctionExpression(
          [],
          //@ts-ignore documentTypeDef will be an object
          j.objectExpression(jsonValueToExpression(documentTypeDef).properties)
        ),
      ])
    ),
  ])

  // Insert after other defineDocument declarations.
  // If none, insert it after imports
  if (imports.length && defineDocument.length) {
    defineDocument[defineDocument.length - 1].insertAfter(ast)
  } else if (imports.length) {
    imports[imports.length - 1].insertAfter(ast)
  } else {
    throw new Error('Expect contentlayer config file to have imports')
  }

  // export default makeSource({
  //   contentDirPath: 'data',
  //   documentTypes: [Blog, Authors],
  // })

  // Add variableName to documentTypes array in makeSource
  const documentTypesField = j(makeSource).find(j.Identifier, { name: 'documentTypes' })

  if (documentTypesField.length) {
    documentTypesField.forEach((p) => {
      const arr = p.parentPath.value.value
      arr.elements.push(variableName)
    })
  } else {
    throw new Error('Expect contentlayer makeSource to have documentType field with an array')
  }

  return program
}
