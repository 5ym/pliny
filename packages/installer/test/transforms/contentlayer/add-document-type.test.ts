import { expect, describe, it } from 'vitest'
import { addDocumentType } from '@pliny/installer'
import type { DocumentTypeDef } from 'contentlayer/source-files'
import type { Options as RecastOptions } from 'recast'
import j from 'jscodeshift'

const recastOptions: RecastOptions = {
  tabWidth: 2,
  arrayBracketSpacing: false,
  objectCurlySpacing: false,
  quote: 'single',
}

describe('contentlayer addDocumentType transform', () => {
  it('should be inserted after other defineDocumentType', () => {
    const file = `
      import { defineDocumentType, makeSource } from 'contentlayer/source-files'
      import rehypeCitation from 'rehype-citation'
  
      const Blog = defineDocumentType(() => ({}))
      export const Author = defineDocumentType(() => ({}))
  
      export default makeSource({
        /* options */
      })
    `
    const newDocumentType: DocumentTypeDef = {
      name: 'Blog',
      filePathPattern: 'blog/**/*.mdx',
      contentType: 'mdx',
      fields: {
        title: { type: 'string', required: true },
        date: { type: 'date', required: true },
      },
    }
    const program = addDocumentType(j(file), 'Blog', newDocumentType)

    expect(program.toSource(recastOptions)).toMatchSnapshot()
  })

  it('should be inserted after imports if no other defineDocumentType', () => {
    const file = `
      import { defineDocumentType, makeSource } from 'contentlayer/source-files'
      import rehypeCitation from 'rehype-citation'
  
      export default makeSource({
        /* options */
      })
    `
    const newDocumentType: DocumentTypeDef = {
      name: 'Blog',
      filePathPattern: 'blog/**/*.mdx',
      contentType: 'mdx',
      fields: {
        title: { type: 'string', required: true },
        date: { type: 'date', required: true },
      },
    }
    const program = addDocumentType(j(file), 'Blog', newDocumentType)

    expect(program.toSource(recastOptions)).toMatchSnapshot()
  })
})
