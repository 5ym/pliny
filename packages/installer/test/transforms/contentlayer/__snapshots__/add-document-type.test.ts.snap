// Vitest Snapshot v1

exports[`contentlayer addDocumentType transform > should be inserted after imports if no other defineDocumentType 1`] = `
"
      import { defineDocumentType, makeSource } from 'contentlayer/source-files'
      import rehypeCitation from 'rehype-citation'

      const Blog = defineDocumentType(() => ({
        'name': 'Blog',
        'filePathPattern': 'blog/**/*.mdx',
        'contentType': 'mdx',

        'fields': {
          'title': {
            'type': 'string',
            'required': true
          },

          'date': {
            'type': 'date',
            'required': true
          }
        }
      }));

      export default makeSource({
        /* options */
      })
    "
`;

exports[`contentlayer addDocumentType transform > should be inserted after other defineDocumentType 1`] = `
"
      import { defineDocumentType, makeSource } from 'contentlayer/source-files'
      import rehypeCitation from 'rehype-citation'

      const Blog = defineDocumentType(() => ({}))
      export const Author = defineDocumentType(() => ({}))

      const Blog = defineDocumentType(() => ({
        'name': 'Blog',
        'filePathPattern': 'blog/**/*.mdx',
        'contentType': 'mdx',

        'fields': {
          'title': {
            'type': 'string',
            'required': true
          },

          'date': {
            'type': 'date',
            'required': true
          }
        }
      }));

      export default makeSource({
        /* options */
      })
    "
`;
